### PR TITLE
[PVR] Remove "group member attributes cached at channel instances" hack

### PR DESCRIPTION
**Pull request by ksooo**
Kodi PVR knows channels, channel group members and channel groups.

A channel represents a TV or radio channel, providing the respective basic attributes, like channel name, channel logo etc.
As every channel can be part of different channel groups, it must not contain attributes which can be different per channel group. Those attributes should be only provided by channel group members. Most prominent example for such a group specific channel attribute is the channel number.
 
IMG_5692

Currently, PVR core uses a strange and hard to understand/maintain construct to provide channel numbers for channels. Instead of providing those via a method CPVRChannelGroupMember::ChannelName(), it has the "concept" (well, bad hack imo) to know a "selected channel group" (assuming there can be only one active group in the GUI at a time) and to copy the channel number and other group specific attributes from the respective group member instances to the corresponding CPVRChannel instances whenever the "selected group" changes. This has been the cause for many hard to track down bugs in the past and there are still some left I know. ;-)

This PR completely removes that silly "caching" logics and strictly uses channel numbers provided by CPVRChannelGroupMember instances. It removes the group specific members and access methods from CPVRChannel. As a side effect it fixes some of the long standing bugs caused by the hack now removed.

BTW: One of my next PRs will remove the concept of "selected channel group", but let's do one step after the other.

I runtime-tested the changes for some weeks now and could not find any regression.

@phunkyfish when you find some time for a review...

### CODE REVIEW

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 1]
```diff
@@ -138,11 +138,18 @@ JSONRPC_STATUS CPVROperations::GetChannelDetails(const std::string &method, ITra
   if (!channelGroupContainer)
     return FailedToExecute;
 
-  std::shared_ptr<CPVRChannel> channel = channelGroupContainer->GetChannelById((int)parameterObject["channelid"].asInteger());
+  std::shared_ptr<CPVRChannel> channel = channelGroupContainer->GetChannelById(
+      static_cast<int>(parameterObject["channelid"].asInteger()));
```
Comment (phunkyfish) [inline xbmc/interfaces/json-rpc/PVROperations.cpp:142]: This is the only part where this doesn't feel right. How do we know that we are retrieving the correct group member?

Probably not an issue if the channelid is group specific but I'm not entirely sure that is the case.

**Review by phunkyfish — COMMENTED** [review]
This is the only part I'm not sure of, everything else looks good. Plus I've tested everything except json rpc and I can't break it!

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/interfaces/json-rpc/PVROperations.cpp:142]: The new code mimics exactly the logic as it was before. It obtains the member of the currently selected channel group or if the requested channel is not member of the selected group, from the all channels group. So, if later on the channel number should be queried from the created item, the exact number as before this PR should be returned.

The 'magic' happens in CServiceBroker::GetPVRManager().GUIActions()->GetChannelGroupMember() called some lines below.

Comment (phunkyfish) [reply xbmc/interfaces/json-rpc/PVROperations.cpp:142]: Cool, OK. Then all good.

**Review by phunkyfish — APPROVED** [review]
LGTM

**Pull request merged**